agent/agent/src/main/java/com/microsoft/applicationinsights/agent/Main.java

```diff
@@ -0,0 +1,158 @@
+    Path agentJar = getAgentJar();
+
+    Path tempJar = Files.createTempFile("applicationinsights-", ".jar");
+
+    byte[] buffer = new byte[1024];
+
+    try (JarInputStream in = new JarInputStream(Files.newInputStream(agentJar))) {
+
+      Manifest inManifest = in.getManifest();
+
+      if (!isSigned(inManifest)) {
+        System.err.println();
+        System.err.println("The agent jar is already unsigned");
+        System.exit(0);
+      }
+
+      Manifest outManifest = new Manifest();
+      outManifest.getMainAttributes().putAll(inManifest.getMainAttributes());
+
+      JarOutputStream out = new JarOutputStream(Files.newOutputStream(tempJar), outManifest);
+
+      JarEntry entry;
+      while ((entry = in.getNextJarEntry()) != null) {
+        if (SIGNATURE_FILE.matcher(entry.getName()).matches()) {
+          continue;
+        }
+        out.putNextEntry(entry);
+        int read;
+        while ((read = in.read(buffer)) != -1) {
+          out.write(buffer, 0, read);
+        }
+      }
+
+      out.close();
+    }
+
+    Path tempClassPath = Files.createTempDirectory("applicationinsights-");
+
+    String resourceName = Copy.class.getName().replace('.', '/') + ".class";
+    InputStream in = Main.class.getClassLoader().getResourceAsStream(resourceName);
+
+    if (in == null) {
+      System.err.println("Could not find resource: " + resourceName);
+      System.exit(1);
+    }
+
+    Path destinationPath = tempClassPath.resolve(resourceName);
+
+    Files.createDirectories(destinationPath.getParent());
+    Files.copy(in, destinationPath);
+
+    List<String> command =
+        asList(
+            getJavaExecutable(),
+            "-cp",
+            tempClassPath.toString(),
+            Copy.class.getName(),
+            tempJar.toString(),
+            agentJar.toString());
+
+    if (DEBUG) {
+      System.out.println(String.join(" ", command));
+    }
+
+    ProcessBuilder builder = new ProcessBuilder().command(command);
+
+    if (DEBUG) {
+      builder.redirectOutput(new File("unsign.out")).redirectError(new File("unsign.err"));
+    }
+
+    builder.start();
+  }
+
+  private static boolean isSigned(Manifest manifest) {
+    for (Map.Entry<String, Attributes> entry : manifest.getEntries().entrySet()) {
+      for (Object key : entry.getValue().keySet()) {
+        if (key instanceof Attributes.Name && key.toString().endsWith("-Digest")) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private static Path getAgentJar() throws URISyntaxException {
+    CodeSource codeSource = Main.class.getProtectionDomain().getCodeSource();
+
+    if (codeSource == null) {
+      throw new IllegalStateException("could not get agent jar location");
+    }
+
+    return Paths.get(codeSource.getLocation().toURI());
+  }
+
+  private static String getJavaExecutable() {
+    return System.getProperty("java.home") + File.separator + "bin" + File.separator + "java";
```

Comment:
Is this always set?

Reply:
this code is gone, it was a fallacy to try to spin up a separate JVM process and overwrite the original agent jar, I mean it worked, but I realized that it could always be a problem if people used that in a script to unsign the jar and copied/moved right after the tool returned, since there was no way (that I could see?) to make the first JVM process not "exit back to the user" until after the second JVM process did it's work